controllers/network/openstackipset_controller.go

```diff
@@ -0,0 +1,264 @@
+				aid.RangeEnd = net.ParseIP(subnet.AllocationRange.AllocationEnd)
+				ipReservation, updatedReservations, err := ipam.AssignIP(aid, net.ParseIP(network.FixedIP))
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+				ipReservation.Gateway = subnet.Gateway
+				ipReservation.Routes = subnet.Routes
+
+				aid.Reservelist = updatedReservations
+				osnet.Spec.Subnets[i].Reservations = append(osnet.Spec.Subnets[i].Reservations, ipReservation)
+
+				err = r.Update(ctx, osnet)
+				if err != nil {
+					return ctrl.Result{}, err
+				}
+
+				instance.Status.IPAddresses[network.Name] = ipReservation
+				instance.Status.Allocated = true
+				break
+			}
+		}
+
+		if controllerutil.AddFinalizer(osnet, fmt.Sprintf("%s-%s", helper.GetFinalizer(), instance.Name)) {
+			err = r.Update(ctx, osnet)
+			if err != nil {
+				r.Log.Info("Failed Adding finalizer")
+				return ctrl.Result{}, err
+			}
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// ReconcileDelete reconciles when resource is deleted
+func (r *OpenStackIPSetReconciler) ReconcileDelete(
+	ctx context.Context,
+	instance *networkv1beta1.OpenStackIPSet,
+	helper *helper.Helper,
+	req ctrl.Request) (ctrl.Result, error) {
+	r.Log.Info("Reconciling IPDataSet delete")
+	networks := instance.Spec.Networks
+
+	for _, network := range networks {
+		osnet := &networkv1beta1.OpenStackNet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      network.Name,
+				Namespace: instance.Namespace,
+			},
+		}
+		err := r.Get(ctx, types.NamespacedName{Name: network.Name, Namespace: instance.Namespace}, osnet)
+		if err != nil {
+			if k8s_errors.IsNotFound(err) {
+				// Request object not found, could have been deleted after reconcile request.
+				// Owned objects are automatically garbage collected.
+				// For additional cleanup logic use finalizers. Return and don't requeue.
+				continue
+			}
+			return ctrl.Result{}, err
+		}
+
+		reservation := instance.Status.IPAddresses[network.Name]
+		// Find the subnet to be updated
+		allSubnetReservations := []networkv1beta1.IPReservation{}
+		subIndex := -1
+		for i, subnet := range osnet.Status.Subnets {
+			if subnet.Name == network.SubnetName {
+				allSubnetReservations = osnet.Status.Subnets[i].Reservations
+				subIndex = i
+				break
+			}
+		}
+
+		// Now delete reservation from allSubnetReservations
+		foundIndex := -1
+		for i, addr := range allSubnetReservations {
+			if addr.IP == reservation.IP {
+				foundIndex = i
+				break
+			}
+		}
+
+		if subIndex >= 0 && foundIndex >= 0 {
+			osnet.Spec.Subnets[subIndex].Reservations = slices.Delete(allSubnetReservations,
+				foundIndex, foundIndex+1)
+		}
+
+		err = r.Update(ctx, osnet)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// Remove the finalizers from networks
+		if controllerutil.RemoveFinalizer(osnet, fmt.Sprintf("%s-%s", helper.GetFinalizer(), instance.Name)) {
+			err := r.Update(ctx, osnet)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+		}
```

Comment:
should change this to only have a single patch call if possible

Reply:
commented earlier.